### PR TITLE
Add the ability to save flamegraphs to a file

### PR DESCRIPTION
**Pull request by CyborgMaster**
This is useful for json api posts, and is my proposal for solving #350.

It follows the same pattern as `flamegraph_embed`, but allows you to save the flamegraph to an external html file, which you can then open in a separate browser tab.

### CODE REVIEW

**Comment by SamSaffron** [comment]
very mixed on this, this would save it on the server and hunting down where the file is, is going to be mega difficult when deployed to a ton of web app servers

**Comment by CyborgMaster** [comment]
This is true, but without this there is no way to look at them even in development, which is what my use case was.  We had a slow endpoint during development and no good way to figure out what it was locally because it was a json endpoint.  This allowed us to dive into what's going on.  I don't think we'd ever try and use it in production.

**Comment by CyborgMaster** [comment]
Any further thoughts?

File: lib/mini_profiler/profiler.rb [hunk 1]
```diff
@@ -331,8 +331,12 @@ def call(env)
       page_struct[:root].record_time((Process.clock_gettime(Process::CLOCK_MONOTONIC) - start) * 1000)
 
       if flamegraph
-        body.close if body.respond_to? :close
-        return client_settings.handle_cookie(self.flamegraph(flamegraph))
+        if query_string =~ /save/
+          ::File.write 'flamegraph.html', flamegraph
```
Comment (SamSaffron) [inline lib/mini_profiler/profiler.rb]: I think you should write this to a file in `tmp` and then output where it was saved eg: 

```
saved flamegraph to tf23.docker  /tmp/flamegraphs_20190201-01023.html`
```

You get the hostname from `hostname`

**Comment by williamcodes** [comment]
Is this PR dead? I'd love the ability to save a flamegraph. I've tried saving it in the browser but it doesn't work properly when I reopen it.

**Comment by SamSaffron** [comment]
I support adding something here, but worried about where it is placing the file. if my comment is addressed I would be happy to merge.

**Comment by CyborgMaster** [comment]
Sorry about that.  Somehow I missed the comment.  I'll make the changes and get it to you soon.

**Comment by CyborgMaster** [comment]
@SamSaffron I saved the files in a temp directory as requested, however I wasn't sure how you wanted the filename outputted.  I can't modify the in-flight request because I don't know it's format; it could be json, xml, protobuf, etc.  In the code right now I just did `puts`, but how do you recommend it be done?

**Comment by SamSaffron** [comment]
> Is this PR dead? I'd love the ability to save a flamegraph. I've tried saving it in the browser but it doesn't work properly when I reopen it.

I think the fundamental issue here is that this is already sort of supported `?pp=flamegraph_embed` allows you to get a flamegraph with all the stuff embedded so you can easily just save the html and load it in another browser / etc. 

What is the specific use case here that goes beyond the ability to share? 

I know I have flip flopped here a lot, but I worry about adding something that will be confusing to explain / use when `flamegraph_embed` already solves the issue.

**Comment by williamcodes** [comment]
@SamSaffron saving the html and loading it in another browser was the first thing I tried. It doesn't save properly. I dunno if it's a css issue or a javascript issue or what. It looks like it might be okay when you first open the file, but the more you try to navigate, the more broken it becomes. If saving as HTML worked, I'd be satisfied with that, I just need to be able to save a file I can attach to a ticket and send to a colleague.

**Comment by SamSaffron** [comment]
Oh it certainly works, as long as you use `?pp=flamegraph_embed` I just retested.

**Comment by CyborgMaster** [comment]
Flamegraph embed is great for when things are opening in a browser.  This solves the problem of when your server just exposes a JSON API that a single page app hits over AJAX.  In that case, I don't believe there is any other way to get ahold of the flamegraph.  Usually this works by injecting it on page load, but that technique cannot work where there is no page being loaded.  More details are in #350.

**Comment by OsamaSayegh** [comment]
Ok, @CyborgMaster explained his use case in https://github.com/MiniProfiler/rack-mini-profiler/issues/350#issuecomment-656761487 and why `?pp=flamegraph`/`?pp=flamegraph_embed` don't suffice for him. Personally I think this PR is OK, but will defer to @SamSaffron to make the final call.

**Comment by nateberkopec** [comment]
This is now possible using the Speedscope visualizer - just use the export button.

**Pull request closed without merging**